xarray/core/concat.py

```diff
@@ -586,9 +642,33 @@ def get_indexes(name):
                     )
                     result_vars[k] = v
             else:
-                combined_var = concat_vars(
-                    vars, dim, positions, combine_attrs=combine_attrs
-                )
+                # do not concat if:
+                # 1. variable is only present in one dataset of multiple datasets and
+                # 2. dim is not a dimension of variable
+                if (
+                    dim not in variables[0].dims
+                    and len(variables) == 1
+                    and len(datasets) > 1
+                ):
+                    combined_var = variables[0]
+                # only concat if variable is in multiple datasets
+                # or if single dataset (GH1988)
+                else:
+                    combined_var = concat_vars(
+                        vars, dim, positions, combine_attrs=combine_attrs
+                    )
+                    # reindex if variable is not present in all datasets
+                    if len(variable_index) < len(concat_index):
+                        try:
+                            fill = fill_value[name]
+                        except (TypeError, KeyError):
+                            fill = fill_value
```

Comment:
Isn't this always a scalar value at present? Do we support a dictionary yet?

Reply:
In the doc-string `fill_value` is given as `scalar` or `dict-like`.

Reply:
Oh I see we use it for align. I think we should convert fill_value to a dictionary always, before the call to align.

Reply:
Great, that removes the need for the try/except and makes this a one liner.

Reply:
I've tried to get this running in https://github.com/pydata/xarray/pull/7400/commits/a823977eb6c69ee90ef7b913de359d074fbd94aa.

Reply:
Great, can't get the worms back into the can. For your suggestion to work this has to be refactored everywhere.

Reply:
I've reverted those changes to the working state.